net.go

```diff
@@ -109,58 +136,81 @@ func DialTimeout(network, addr string, timeout time.Duration, ctx *Ctx,
 func DialSession(network, addr string, ctx *Ctx, flags DialFlags,
 	session []byte) (*Conn, error) {
 	var d net.Dialer
-	return dialSession(d, network, addr, ctx, flags, session)
-}
-
-func dialSession(d net.Dialer, network, addr string, ctx *Ctx, flags DialFlags,
-	session []byte) (*Conn, error) {
-	host, _, err := net.SplitHostPort(addr)
+	if err := prepareCtx(ctx); err != nil {
+		return nil, err
+	}
+	conn, err := createConnection(d, context.Background(), network, addr)
 	if err != nil {
 		return nil, err
 	}
+	return createSession(conn, flags, addr, ctx, session)
+}
+
+func prepareCtx(ctx *Ctx) error {
 	if ctx == nil {
 		var err error
 		ctx, err = NewCtx()
 		if err != nil {
-			return nil, err
+			return err
 		}
 		// TODO: use operating system default certificate chain?
 	}
+	return nil
+}
 
-	c, err := d.Dial(network, addr)
+func createConnection(d net.Dialer, context context.Context, network,
+	addr string) (net.Conn, error) {
+	c, err := d.DialContext(context, network, addr)
 	if err != nil {
 		return nil, err
 	}
-	conn, err := Client(c, ctx)
-	if err != nil {
-		c.Close()
-		return nil, err
-	}
-	if session != nil {
-		err := conn.setSession(session)
-		if err != nil {
-			c.Close()
-			return nil, err
-		}
-	}
+	return c, nil
+}
+
+func applyFlags(conn *Conn, host string, flags DialFlags) error {
+	var err error
 	if flags&DisableSNI == 0 {
 		err = conn.SetTlsExtHostName(host)
 		if err != nil {
 			conn.Close()
```

Comment:
We open the connection out of the scope of the function. So let and close it out of the scope in the error case.

Reply:
Updated.